File: 001-500/0244. Shortest Word Distance II.java
// Medium
// Hash Table, Two Pointers
// https://leetcode.cn/problems/shortest-word-distance-ii/

import java.util.*;

class WordDistance {
  Map<String, List<Integer>> map;

  // 構造函數可以用哈希表
  public WordDistance(String[] wordsDict) {
      map = new HashMap<>();
      for (int i = 0; i < wordsDict.length; i++) {
          String word = wordsDict[i];
          if (!map.containsKey(word)) {
              map.put(word, new ArrayList<>());
          }
          map.get(word).add(i);
      }
  }
  
  // 找最短的單詞路徑可以用雙指針
  public int shortest(String word1, String word2) {
      List<Integer> list1 = map.get(word1);
      List<Integer> list2 = map.get(word2);
      int minDis = Integer.MAX_VALUE;
      int i = 0, j = 0;

      while (i < list1.size() && j < list2.size()) {
          int index1 = list1.get(i), index2 = list2.get(j);
          if (index1 < index2) {
              minDis = Math.min(minDis, index2 - index1);
              i++;
          } else {
              minDis = Math.min(minDis, index1 - index2);
              j++;
          }
      }
      return minDis;
  }
}

/**
 * 最短單詞距離：設計一個類來接受一系列單詞，能夠多次有效地計算指定兩個單詞之間的最短距離
 **/

File: 0001-0500/0244. Shortest Word Distance II.py
from typing import List

class WordDistance:

    def __init__(self, wordsDict: List[str]):
        self.word_map = {}

        for i, word in enumerate(wordsDict):
            if word not in self.word_map:
                self.word_map[word] = []
            self.word_map[word].append(i)
    
    def shortest(self, word1: str, word2: str) -> int:
        idx1 = self.word_map[word1]
        idx2 = self.word_map[word2]

        i = j = 0
        min_dist = float('inf')
        
        while i < len(idx1) and j < len(idx2):
            min_dist = min(min_dist, abs(idx1[i] - idx2[j]))

            if idx1[i] < idx2[j]: 
                i += 1
            else:
                j += 1

        return min_dist
    